library/std/src/os/linux/process.rs

```diff
@@ -0,0 +1,47 @@
+//! Linux-specific extensions to primitives in the `std::process` module.
+
+#![unstable(feature = "linux_pidfd", issue = "none")]
+
+use crate::process;
+use crate::sys_common::AsInnerMut;
+use crate::io::Result;
+
+/// Os-specific extensions to [`process::Child`]
+///
+/// [`process::Child`]: crate::process::Child
+pub trait ChildExt {
+    /// Obtains the pidfd created for this child process, if available.
+    ///
+    /// A pidfd will only ever be available if `create_pidfd(true)` was called
+    /// when the corresponding `Command` was created.
+    ///
+    /// Even if `create_pidfd(true)` is called, a pidfd may not be available
+    /// due to an older version of Linux being in use, or if
+    /// some other error occured.
+    ///
+    /// See `man pidfd_open` for more details about pidfds.
+    fn pidfd(&self) -> Result<i32>;
+}
+
+/// Os-specific extensions to [`process::Command`]
+///
+/// [`process::Command`]: crate::process::Command
+pub trait CommandExt {
+    /// Sets whether or this `Command` will attempt to create a pidfd
+    /// for the child. If this method is never called, a pidfd will
+    /// not be crated.
+    ///
+    /// The pidfd can be retrieved from the child via [`ChildExt::pidfd`]
+    ///
+    /// A pidfd will only be created if it is possible to do so
+    /// in a guaranteed race-free manner (e.g. if the `clone3` system call is
+    /// supported). Otherwise, [`ChildExit::pidfd`] will return an error.
+    fn create_pidfd(&mut self, val: bool) -> &mut process::Command;
```

Comment:
Why does the user need to request it explicitly? As long as clone3 is available is there any cost to always obtaining it?

Reply:
Always opening a pidfd would cause it to be leaked, since we don't ever close it in the standard library. This would eventually exhaust all of the file descriptors for the process with leaked pidfds.

Reply:
Well, it could be closed when the `Child` is dropped and  one would have to `take` it out if one wants to keep using it, but yeah that would indeed be an extra cost.

Reply:
I think ownership and moving would be appropriate here since someone has to be responsible for closing it and if one can retrieve copies it becomes unclear who is responsible.

Reply:
1. Without seeing this discussion, I would presume that Process owns the pidfd and will close it on drop.
2. The pidfd should be closed when clone3 succeeds but exec fails, since a user has no ability to do so on their own, right?

Reply:
> Without seeing this discussion, I would presume that Process owns the pidfd and will close it on drop.

If we do that, then I think we should add a way to prevent this behavior (other then calling `mem::forget` on the `Child`). Otherwise, users who just need the pidfd will be forced to potentially leak memory (if `Child` ever heap allocates internally), or keep a `Child` around when it's not actually needed.

> The pidfd should be closed when clone3 succeeds but exec fails, since a user has no ability to do so on their own, right?

The pidfd only exists in the parent process. Once `clone3` succeeds, the pidfd is perfectly useable, regardless of what happens in the child process from that point on.

Reply:
> The pidfd only exists in the parent process. Once clone3 succeeds, the pidfd is perfectly useable, regardless of what happens in the child process from that point on.

My point was that spawn returns an error in that case, so the user cannot access pidfd.

Reply:
> If we do that, then I think we should add a way to prevent this behavior (other then calling mem::forget on the Child). 

To keep things RAII we should introduce a new wrapper type akin to other fd-wrapping IO types such as `File`, `TcpStream` and so on. E.g. `struct ProcessDescriptor(RawFd)` which could implement `Drop`.

Then `Child` could have
* `take_pidfd(&mut self) -> ProcessDescriptor` 
* `pidfd(&self) -> &ProcessDescriptor`

`ProcessDescriptor` then could implement `IntoRawFd` and `AsRawFd`

Reply:
Oh, that's a good point. I'll fix that.